app/Http/Controllers/Backend/CategoryController.php

```diff
@@ -0,0 +1,133 @@
+        $data["cates"] = Category::tree();
+        return view('backend.categories.create')->with($data);
+    }
+
+    /**
+     * Store a newly created resource in storage.
+     *
+     * @param \Illuminate\Http\Request $request request
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function store(CategoryRequest $request)
+    {
+        $addcate = new Category;
+        $addcate->cate_name = $request->cate_name;
+        $addcate->cate_description = $request->cate_description;
+        $addcate->cate_status = $request->cate_status;
+        $addcate->parent_id = $request->parent_id;
+        if ($request->hasFile('cate_image')) {
+            $fileName = $request->file('cate_image')->getClientOriginalName();
+            $addcate->cate_image = $fileName;
+            $request->file('cate_image')->move(public_path(config('app.upload')), $fileName);
+        }
+        $addcate->save();
+        $request->session()->flash('message', trans('messages.categoryadd'));
+        return redirect()->route('admin.categories.index');
+    }
+
+    /**
+     * Show the form for editing the specified resource.
+     *
+     * @param int $id id
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function edit($id)
+    {
+        $data["data"] = Category::findOrFail($id);
+        $data["cates"] = Category::tree();
+        return view('backend.categories.edit')->with($data);
+    }
+
+    /**
+     * Update the specified resource in storage.
+     *
+     * @param \Illuminate\Http\Request $request CategoryRequest
+     * @param int                      $id      id
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function update(CategoryRequest $request, $id)
+    {
+        $editcate = Category::findOrFail($id);
+        $editcate->cate_name = $request->cate_name;
+        $editcate->cate_description = $request->cate_description;
+        $editcate->cate_status = $request->cate_status;
+        $editcate->parent_id = $request->parent_id;
+        if ($request->hasFile('cate_image')) {
+            if ($editcate->cate_image) {
+                file::delete(public_path('upload/') . $editcate->cate_image);
+            }
+            $fileName = $request->file('cate_image')->getClientOriginalName();
+            $editcate->cate_image = $fileName;
+            $request->file('cate_image')->move(public_path(config('app.upload')), $fileName);
+        }
+        $editcate->save();
+        $request->session()->flash('message', trans('messages.categoryupdate'));
+        return redirect()->route('admin.categories.index');
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     *
+     * @param int $id id
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function destroy($id)
+    {
+        $delcate = Category::findOrFail($id);
+        $cate = Category::all();
+        $stemp = 0;
+        foreach ($cate as $item) {
```

Comment:
You want to check if the category have children category, then don't allow to delete, right ?
If it's true, this code is not good. Why don't you define a relationship between parent and children. Then it's very easy to check.